polkadot/node/core/provisioner/src/lib.rs

```diff
@@ -643,60 +652,119 @@ async fn request_backable_candidates(
 ) -> Result<Vec<(CandidateHash, Hash)>, Error> {
 	let block_number = get_block_number_under_construction(relay_parent, sender).await?;
 
-	let mut selected_candidates = Vec::with_capacity(availability_cores.len());
+	// These need to be ordered by core index. The runtime makes this assertion.
+	let mut selected_candidates: BTreeMap<CoreIndex, (CandidateHash, Hash)> = BTreeMap::new();
+
+	// Record which cores are scheduled for each paraid. Use a BTreeMap because
+	// we'll need to iterate through them.
+	let mut scheduled_cores: BTreeMap<ParaId, HashSet<CoreIndex>> = BTreeMap::new();
+	// The on-chain ancestors of a para present in availability-cores.
+	let mut ancestors: HashMap<ParaId, Ancestors> =
+		HashMap::with_capacity(availability_cores.len());
 
 	for (core_idx, core) in availability_cores.iter().enumerate() {
-		let (para_id, required_path) = match core {
+		let core_idx = CoreIndex(core_idx as u32);
+		match core {
 			CoreState::Scheduled(scheduled_core) => {
-				// The core is free, pick the first eligible candidate from
-				// the fragment tree.
-				(scheduled_core.para_id, Vec::new())
+				scheduled_cores
+					.entry(scheduled_core.para_id)
+					.or_insert(HashSet::new())
+					.insert(core_idx);
 			},
 			CoreState::Occupied(occupied_core) => {
-				if bitfields_indicate_availability(core_idx, bitfields, &occupied_core.availability)
-				{
+				let is_available = bitfields_indicate_availability(
+					core_idx.0 as usize,
+					bitfields,
+					&occupied_core.availability,
+				);
+
+				if is_available {
+					ancestors
+						.entry(occupied_core.para_id())
+						.or_default()
+						.entry(occupied_core.candidate_hash)
+						.or_default()
+						.record();
+
 					if let Some(ref scheduled_core) = occupied_core.next_up_on_available {
-						// The candidate occupying the core is available, choose its
-						// child in the fragment tree.
-						//
-						// TODO: doesn't work for on-demand parachains. We lean hard on the
-						// assumption that cores are fixed to specific parachains within a session.
-						// https://github.com/paritytech/polkadot/issues/5492
-						(scheduled_core.para_id, vec![occupied_core.candidate_hash])
-					} else {
-						continue
+						// Request a new backable candidate for the newly scheduled para id.
+						scheduled_cores
+							.entry(scheduled_core.para_id)
+							.or_insert(HashSet::new())
+							.insert(core_idx);
 					}
+				} else if occupied_core.time_out_at > block_number {
+					// Not timed out and not available.
+					ancestors
```

Comment:
Point 2:

> Instead of setting timeout to true, we could just not add the timedout candidate to the ancestry to begin with. (voiding my argument above that we can lift the adding to ancestry up.) Then in fragment_tree, there would be a hole in the chain - which is fine, we just need to stop once we find no matching child - ignoring any remaining elements in ancestors. (Should no longer be an error.)

Yes, we could do that. I didn't do this because I figured we'd want to distinguish between a hole in the chain given by a timeout or a hole in the chain due to a runtime error backing candidates which don't form a chain or incorrectly freeing the cores.

So if the cores were occupied by A->B->C and the runtime freed B's core, we'd end up treating B as timed out and suggest B again as a new backable candidate. If we can distinguish between these two situations, we won't supply any more candidates to the runtime until it recovers from the error.

Reply:
> So if the cores were occupied by A->B->C and the runtime freed B's core, we'd end up treating B as timed out and suggest B again as a new backable candidate. If we can distinguish between these two situations, we won't supply 

the runtime can not free `B` without also freeing `A` ... so there would not be a hole in that case. I am not against defensive programming, but we should not complicate node code to enhance performance in case the runtime has a bug.

> If the count is larger than 1, we can be sure there's a cycle, 

Ok. In theory this is really possible. The candidate descriptor contains the relay parent so same candidates would be limited to be anchored to the same relay parent, but with asynchronous backing and also elastic scaling there is nothing preventing a parachain from not being a chain and producing the same candidate twice and then later moving on. This can only be prevented by adding the parent candidate hash to the descriptor itself. But for all intents and purposes those candidates are already exactly the same. They take the same parent head data and they produce the exact same output with the exact same "block" data. So things like this could exist:

A <- B <- A

now if we keep building, we could get:

A <- B <- A <- C

but given that the second C arrives at exactly the same output and it can not have consumed or sent any messages (otherwise they would not be the same), hence from Polkadot's perspective this is the same as:

A <- C

So if the cycle A <- B <- A exists within a single block (elastic scaling), we can just drop the second A in the runtime - or the first, it does not matter (they are the same). In fact the first `A` must also depend on `B` or more precisely the head data produced by B. 

I would say for now, we should really just drop duplicates and in the next iteration we absolutely should have the parent candidate hash in the descriptor to stop this madness altogether.

Reply:
> I would say for now, we should really just drop duplicates and in the next iteration we absolutely should have the parent candidate hash in the descriptor to stop this madness altogether.

Sounds like a good idea

> the runtime can not free B without also freeing A ... so there would not be a hole in that case. I am not against defensive programming, but we should not complicate node code to enhance performance in case the runtime has a bug.

right, I agree.

So, I'll remove the `AncestorState` and just use a HashSet to express the ancestors. Should be much simpler. Great suggestion

Reply:
done, PTAL